File: android/src/main/java/com/bridge/WebEvents.kt
package com.bridge

import android.webkit.WebView
import org.json.JSONObject

/** Built-in web → native event names (`NativeBridge.send`). */
object WebEvents {
    const val WEBVIEW_LOADED = "WEBVIEW_LOADED"
}

enum class WebViewLoadedPhase(val raw: String) {
    DOM("dom"),
    COMPLETE("complete"),
    MANUAL("manual"),
    ;

    companion object {
        fun from(raw: String?): WebViewLoadedPhase =
            entries.find { it.raw == raw } ?: MANUAL
    }
}

/** Parsed payload for [WebEvents.WEBVIEW_LOADED]. */
data class WebViewLoadedPayload(
    val url: String,
    val title: String,
    val timestamp: Long,
    val readyState: String,
    val phase: WebViewLoadedPhase,
    val referrer: String?,
    val webViewId: String?,
    val route: String?,
    val extras: Map<String, Any?>,
) {
    companion object {
        private val KNOWN_KEYS = setOf(
            "event", "url", "title", "timestamp", "readyState",
            "phase", "referrer", "webViewId", "route",
        )

        fun parse(payload: Any?, fallbackWebViewId: String): WebViewLoadedPayload {
            val json = when (payload) {
                is JSONObject -> payload
                is Map<*, *> -> JSONObject(payload)
                else -> JSONObject()
            }
            val extras = mutableMapOf<String, Any?>()
            json.keys().forEach { key ->
                if (key !in KNOWN_KEYS) extras[key] = json.opt(key)
            }
            return WebViewLoadedPayload(
                url = json.optString("url", ""),
                title = json.optString("title", ""),
                timestamp = json.optLong("timestamp", System.currentTimeMillis()),
                readyState = json.optString("readyState", "unknown"),
                phase = WebViewLoadedPhase.from(json.optString("phase")),
                referrer = json.optString("referrer").takeIf { it.isNotEmpty() },
                webViewId = json.optString("webViewId").takeIf { it.isNotEmpty() }
                    ?: fallbackWebViewId,
                route = json.optString("route").takeIf { it.isNotEmpty() },
                extras = extras,
            )
        }
    }
}

typealias WebViewLoadedListener = (
    payload: WebViewLoadedPayload,
    webViewId: String,
    webView: WebView,
) -> Unit

File: ios/Sources/NativeWebViewBridge/WebEvents.swift
import Foundation
import WebKit

/// Built-in web → native event names (`NativeBridge.send`).
public enum WebEvents {
    public static let webViewLoaded = "WEBVIEW_LOADED"
}

public enum WebViewLoadedPhase: String {
    case dom
    case complete
    case manual

    public static func from(_ raw: String?) -> WebViewLoadedPhase {
        guard let raw = raw, let phase = WebViewLoadedPhase(rawValue: raw) else { return .manual }
        return phase
    }
}

/// Parsed payload for [WebEvents.webViewLoaded].
public struct WebViewLoadedPayload {
    public let url: String
    public let title: String
    public let timestamp: TimeInterval
    public let readyState: String
    public let phase: WebViewLoadedPhase
    public let referrer: String?
    public let webViewId: String?
    public let route: String?
    public let extras: [String: Any]

    public static func parse(_ payload: Any?, fallbackWebViewId: String) -> WebViewLoadedPayload {
        let dict: [String: Any]
        if let map = payload as? [String: Any] {
            dict = map
        } else {
            dict = [:]
        }

        let known: Set<String> = [
            "event", "url", "title", "timestamp", "readyState",
            "phase", "referrer", "webViewId", "route",
        ]
        var extras: [String: Any] = [:]
        for (key, value) in dict where !known.contains(key) {
            extras[key] = value
        }

        let webViewId = (dict["webViewId"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            ?? fallbackWebViewId

        return WebViewLoadedPayload(
            url: dict["url"] as? String ?? "",
            title: dict["title"] as? String ?? "",
            timestamp: (dict["timestamp"] as? NSNumber)?.doubleValue ?? Date().timeIntervalSince1970 * 1000,
            readyState: dict["readyState"] as? String ?? "unknown",
            phase: WebViewLoadedPhase.from(dict["phase"] as? String),
            referrer: (dict["referrer"] as? String).flatMap { $0.isEmpty ? nil : $0 },
            webViewId: webViewId,
            route: (dict["route"] as? String).flatMap { $0.isEmpty ? nil : $0 },
            extras: extras
        )
    }
}

public typealias WebViewLoadedHandler = (
    _ payload: WebViewLoadedPayload,
    _ webViewId: String,
    _ webView: WKWebView
) -> Void
